util/clustering/kmeans_clustering.py

```diff
@@ -0,0 +1,231 @@
+        # img = load_img(BytesIO(object.get()['Body'].read()), target_size=(224,224))
+        img = load_img(file, target_size=(224,224))
+
+        img = np.array(img) 
+        # reshape the data for the model reshape(num_of_samples, dim 1, dim 2, channels)
+        reshaped_img = img.reshape(1,224,224,3) 
+        # prepare image for model
+        imgx = preprocess_input(reshaped_img)
+        # get the feature vector
+        features = model.predict(imgx, use_multiprocessing=True)
+        return features
+    
+    data = {}
+    # p = r"CHANGE TO A LOCATION TO SAVE FEATURE VECTORS"
+
+    # loop through each image in the dataset
+    # temporarily trying first 1000 images
+    for i in tqdm(range(1000)):
+        # try to extract the features and update the dictionary
+        feat = extract_features(detection_train_images[i],model)
+        data[detection_train_images[i]] = feat
+        
+        # if something fails, save the extracted features as a pickle file (optional)
+        # except:
+        #     with open(p,'wb') as file:
+        #         pickle.dump(data,file)
+            
+    
+    # get a list of the filenames
+    filenames = np.array(list(data.keys()))
+
+    # get a list of just the features
+    feat = np.array(list(data.values()))
+
+    # reshape so that there are 210 samples of 4096 vectors
+    feat = feat.reshape(-1,4096)
+
+
+    # # reduce the amount of dimensions in the feature vector
+    print('------------ feature reduction ------------')
+    pca = PCA(n_components=100, random_state=22)
+    x = pca.fit_transform(feat)
+
+    os.chdir(file_dir)
+
+    # this is just incase you want to see which value for k might be the best 
+    # note: this didn't really work very well (gave best number of clusters = 5
+    # which is obviously too little)
+    # sse = []
+    # list_k = list(range(5, 1000))
+    # max_diff = 0
+    # prev_sse = 0
+    # best_k = 0
+    # attempts = 0
+
+    # print('------------ checking clusters ------------')
+
+    # find the number of clusters that produces the greatest change in sse. 
+    # for k in tqdm(list_k):
+    #     km = KMeans(n_clusters=k, random_state=22)
+    #     km.fit(x)
+    #     inertia = km.inertia_
+    #     if abs(inertia - prev_sse) > max_diff:
+    #         print('new best k : ' + str(k))
+    #         max_diff = inertia 
+    #         best_k = k 
+    #         attempts = 0
+    #     prev_see = inertia
+    #     sse.append(inertia)
+    #     attempts += 1
+
+    # print('best number of clusters: ' + str(best_k))
+
+    # # Plot sse against k
+    # print('------------ plotting sse against k ------------')
+    # plt.figure(figsize=(6, 6))
+    # plt.plot(list_k, sse)
+    # plt.xlabel(r'Number of clusters *k*')
+    # plt.ylabel('Sum of squared distance')
+    # plt.show()
+    # plt.savefig(args.output_data_dir + '/cluster_analysis.png')
+
+    # num_clusters = best_k
```

Comment:
If the comments are left here for testing purposes, then can you move them into separate functions? They make the code hard to read.

If they aren't for testing or any other specific purpose (e.g. "idk if I need this later" is a good specific purpose) then just remove them.